File: sp/java/pingpong/core/Packet.java
package pingpong.core;

import java.util.*;
import java.util.stream.Collectors;
import org.json.JSONObject; // Assumes org.json library

public class Packet {
    public Channel channel;
    public Status status;
    public Peering peering;
    public Transformer transformer;
    public String separator = "\n";

    public Packet(Channel channel, Status status, Peering peering, Transformer transformer) {
        this.channel = channel;
        this.status = status;
        this.peering = peering;
        this.transformer = transformer;
    }

    @Override
    public String toString() {
        return String.join(separator, 
            channel.toString(), 
            status.toString(), 
            peering.toString(), 
            transformer.toString()
        );
    }

    public static Packet fromString(String data) {
        return fromString(data, "\n");
    }

    public static Packet fromString(String data, String separator) {
        String[] parts = data.split(separator);
        // Transformer takes the remainder of the parts
        List<String> transformerParts = Arrays.asList(parts).subutList(3, parts.length);
        
        return new Packet(
            Channel.fromString(parts[0]),
            Status.fromString(parts[1]),
            Peering.fromString(parts[2]),
            Transformer.fromStrings(transformerParts)
        );
    }

    public static Packet fromSP(String json) {
        JSONObject data = new JSONObject(json);

        Channel channel = Channel.fromString(data.getString("channel"));
        Status status = Status.fromString("0|0000");
        Peering peering = Peering.fromMap(data.getJSONObject("peering").toMap());
        Transformer transformer = Transformer.fromString(data.getString("transformer"));

        return new Packet(channel, status, peering, transformer);
    }
}

class Channel {
    public int path;
    public int layer;
    public int signal;

    public Channel(int path, int layer, int signal) {
        this.path = path;
        this.layer = layer;
        this.signal = signal;
    }

    public boolean isSignal() { return this.signal == 1; }
    public void setSignal(boolean value) { this.signal = value ? 1 : 0; }

    public boolean above(int layer) { return this.layer > layer; }

    @Override
    public String toString() {
        return path + "|" + layer + "|" + signal;
    }

    public static Channel fromString(String data) {
        String[] parts = data.split("\\|");
        return new Channel(
            Integer.parseInt(parts[0]), 
            Integer.parseInt(parts[1]), 
            Integer.parseInt(parts[2])
        );
    }
}

class Status {
    public int age;
    public String health;

    public Status(int age, String health) {
        this.age = age;
        this.health = health;
    }

    public boolean tooOld(int max) { return this.age > max; }
    public boolean ill(String pattern) { return !this.health.equals(pattern); }
    public void getOlder(int inc) { this.age += inc; }

    @Override
    public String toString() {
        return age + "|" + health;
    }

    public static Status fromString(String data) {
        String[] parts = data.split("\\|");
        return new Status(Integer.parseInt(parts[0]), parts[1]);
    }
}

class Peering {
    public long timestamp;
    public String from;
    public String to;
    public String subject;

    public Peering(String from, String to, String subject) {
        this(from, to, subject, System.currentTimeMillis() * 1000);
    }

    public Peering(String from, String to, String subject, long timestamp) {
        this.from = from;
        this.to = to;
        this.subject = subject;
        this.timestamp = timestamp;
    }

    @Override
    public String toString() {
        return timestamp + " " + from + " " + to + " " + subject;
    }

    public static Peering fromString(String data) {
        String[] parts = data.split(" ");
        return new Peering(parts[1], parts[2], parts[3], Long.parseLong(parts[0]));
    }

    public static Peering fromMap(Map<String, Object> data) {
        return new Peering(
            (String) data.get("from"), 
            (String) data.get("to"), 
            (String) data.get("subject")
        );
    }
}

class Transformer {
    public int pointer;
    public byte[] operators;
    public String data;

    public Transformer(int pointer, byte[] operators, String data) {
        this.pointer = pointer;
        this.operators = operators;
        this.data = data;
    }

    public int shift(int delta) {
        int p = this.pointer + delta;
        if (p >= 0 && p < this.operators.length) {
            this.pointer = p;
        }
        return this.pointer;
    }

    public void apply() { shift(1); }

    @Override
    public String toString() {
        String ops = "";
        for (int i = 0; i < operators.length; i++) {
            ops += operators[i] + (i == operators.length - 1 ? "" : ",");
        }
        return pointer + "|" + ops + "\n" + data;
    }

    public static Transformer fromStrings(List<String> list) {
        String meta = list.get(0);
        String data = list.stream().skip(1).collect(Collectors.joining("\n"));
        
        String[] metaParts = meta.split("\\|");
        int pointer = Integer.parseInt(metaParts[0]);
        String[] opsParts = metaParts[1].split(",");
        byte[] operators = new byte[opsParts.length];
        for (int i = 0; i < opsParts.length; i++) {
            operators[i] = Byte.parseByte(opsParts[i]);
        }
        
        return new Transformer(pointer, operators, data);
    }

    public static Transformer fromString(String str) {
        return fromStrings(Arrays.asList(str.split("\n")));
    }
}

File: sp/rust/pingpong/core/packet.rs
use std::fmt;
use std::str::FromStr;
use serde_json::{Value, json}; // Requires serde_json crate

pub struct Packet {
    pub channel: Channel,
    pub status: Status,
    pub peering: Peering,
    pub transformer: Transformer,
    pub separator: String,
}

impl Packet {
    pub fn new(channel: Channel, status: Status, peering: Peering, transformer: Transformer) -> Self {
        Self {
            channel,
            status,
            peering,
            transformer,
            separator: "\n".to_string(),
        }
    }

    /// Equivalent to JS fromSP(json)
    pub fn from_sp(json_str: &str) -> Result<Self, Box<dyn std::error::Error>> {
        let v: Value = serde_json::from_str(json_str)?;
        
        let channel = Channel::from_str(v["channel"].as_str().unwrap_or("0|0|0"))?;
        let status = Status::from_str("0|0000")?;
        let peering = Peering::from_value(&v["peering"])?;
        let transformer = Transformer::from_str(v["transformer"].as_str().unwrap_or("0|0\n"))?;

        Ok(Packet::new(channel, status, peering, transformer))
    }
}

impl fmt::Display for Packet {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}{}{}{}{}{}{}",
            self.channel, self.separator, 
            self.status, self.separator, 
            self.peering, self.separator, 
            self.transformer
        )
    }
}

impl FromStr for Packet {
    type Err = Box<dyn std::error::Error>;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let parts: Vec<&str> = s.split('\n').collect();
        if parts.len() < 4 {
            return Err("Invalid packet format".into());
        }

        Ok(Packet::new(
            Channel::from_str(parts[0])?,
            Status::from_str(parts[1])?,
            Peering::from_str(parts[2])?,
            Transformer::from_strings(&parts[3..].iter().map(|&x| x.to_string()).collect())?,
        ))
    }
}

pub struct Channel {
    pub path: i32,
    pub layer: i32,
    pub signal: i32,
}

impl Channel {
    pub fn is_signal(&self) -> bool {
        self.signal == 1
    }

    pub fn set_is_signal(&mut self, value: bool) {
        self.signal = if value { 1 } else { 0 };
    }

    pub fn above(&self, layer: i32) -> bool {
        self.layer > layer
    }
}

impl fmt::Display for Channel {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}|{}|{}", self.path, self.layer, self.signal)
    }
}

impl FromStr for Channel {
    type Err = Box<dyn std::error::Error>;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let parts: Vec<&str> = s.split('|').collect();
        Ok(Channel {
            path: parts[0].parse()?,
            layer: parts[1].parse()?,
            signal: parts[2].parse()?,
        })
    }
}

pub struct Status {
    pub age: i32,
    pub health: String,
}

impl Status {
    pub fn too_old(&self, max: i32) -> bool {
        self.age > max
    }

    pub fn ill(&self, pattern: &str) -> bool {
        self.health != pattern
    }

    pub fn get_older(&mut self, inc: i32) {
        self.age += inc;
    }
}

impl fmt::Display for Status {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}|{}", self.age, self.health)
    }
}

impl FromStr for Status {
    type Err = Box<dyn std::error::Error>;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let parts: Vec<&str> = s.split('|').collect();
        Ok(Status {
            age: parts[0].parse()?,
            health: parts[1].to_string(),
        })
    }
}

pub struct Peering {
    pub timestamp: u64,
    pub from: String,
    pub to: String,
    pub subject: String,
}

impl Peering {
    pub fn from_value(v: &Value) -> Result<Self, Box<dyn std::error::Error>> {
        Ok(Peering {
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)?
                .as_micros() as u64,
            from: v["from"].as_str().unwrap_or("").to_string(),
            to: v["to"].as_str().unwrap_or("").to_string(),
            subject: v["subject"].as_str().unwrap_or("").to_string(),
        })
    }
}

impl fmt::Display for Peering {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} {} {} {}", self.timestamp, self.from, self.to, self.subject)
    }
}

impl FromStr for Peering {
    type Err = Box<dyn std::error::Error>;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let parts: Vec<&str> = s.split(' ').collect();
        Ok(Peering {
            timestamp: parts[0].parse()?,
            from: parts[1].to_string(),
            to: parts[2].to_string(),
            subject: parts[3].to_string(),
        })
    }
}

pub struct Transformer {
    pub pointer: usize,
    pub operators: Vec<u8>,
    pub data: String,
}

impl Transformer {
    pub fn shift(&mut self, delta: isize) -> usize {
        let p = self.pointer as isize + delta;
        if p >= 0 && (p as usize) < self.operators.len() {
            self.pointer = p as usize;
        }
        self.pointer
    }

    pub fn apply(&mut self) {
        self.shift(1);
    }

    pub fn from_strings(list: &Vec<String>) -> Result<Self, Box<dyn std::error::Error>> {
        let meta = &list[0];
        let meta_parts: Vec<&str> = meta.split('|').collect();
        let pointer = meta_parts[0].parse()?;
        let operators = meta_parts[1]
            .split(',')
            .map(|s| s.parse::<u8>().unwrap_or(0))
            .collect();
        let data = list[1..].join("\n");

        Ok(Transformer { pointer, operators, data })
    }
}

impl fmt::Display for Transformer {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let ops: Vec<String> = self.operators.iter().map(|b| b.to_string()).collect();
        write!(f, "{}|{}\n{}", self.pointer, ops.join(","), self.data)
    }
}

impl FromStr for Transformer {
    type Err = Box<dyn std::error::Error>;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let lines: Vec<String> = s.split('\n').map(|l| l.to_string()).collect();
        Self::from_strings(&lines)
    }
}